### PR TITLE
[FIX] account_reports: 'domain' engine: resolve calls to ref() when loading formula shortcuts

### PR DESCRIPTION
**Pull request by oco-odoo**
The full syntax should make use of eval="..." directly in XML. Doing this leads to the fact the formulas, when displayed in the UI, don't contain any call to ref(), but directly the ids obtained from these calls.

Before this commit, the shortcut syntax used to declare the same expressions did not work in the same way and kept an explicit ref() in the formula to evaluate. We now do the same in both places.

We don't want to support explicit ref() in formulas (a user modifying a formula will anyway not know the xmlid to call). The enterprise part of this PR takes care of that.

### CODE REVIEW

**Comment by xmo-odoo** [comment]
The explanations confuse me. From what I can see on the enterprise side, the formula strings could contain `ref()` calls which were not evaluated. Now the formula goes through an `eval` meaning the `ref` call is evaluated immediately, and the result (a python literal) is implicitly stringified when stored into the `char`) field.

So what is this change for? Fixing up `domain_formula` because the pseudo-functions mean the same solution can't be applied? If that is the case, that's what should be explained in the commit message, because from its explanations it looks like odoo/enterprise#32161 is all that's needed.

Also I have absolutely no idea what 

> Before this commit, the shortcut syntax used to declare the same expressions did not work in the same way in kept an explicit ref() in the formula to evaluate. We now do the same in both places.

is saying. What is "the shortcut syntax"? What "same expressions"? Are there missing words?

> We don't want to support explicit ref() in formulas (a user modifying a formula will anyway not know the xmlid to call).

Would they really be more cognisant of the raw database id to call?

**Comment by oco-odoo** [comment]
> Also I have absolutely no idea what
> 
> > Before this commit, the shortcut syntax used to declare the same expressions did not work in the same way in kept an explicit ref() in the formula to evaluate. We now do the same in both places.
> 
> is saying. What is "the shortcut syntax"? What "same expressions"? Are there missing words?

Well, that's why you don't see why enterprise isn't enough. Maybe ask this first ?
There is no missing word, just a typo "in the same way in kept an explicit ref() " => "in the same way and kept an explicit ref()" ; I'll fix that. 

Anyway, I can explain, but don't complain if the commit message considers you know the basic models used in account_reports... I'll let accounting reviewers decide whether more explanations about this are required in the commit message or not. 

The "shortcut syntax" is a way to declare expressions in XML with less lines. 

Full declaration example: https://github.com/odoo/odoo/blob/16.0/addons/l10n_be/data/account_tax_report_data.xml#L37

Example with a shortcut : https://github.com/odoo/enterprise/blob/16.0/account_reports/data/balance_sheet.xml#L40

Shortcuts are just a shorter way to create the same expressions. So, it would make sense that loading them from the *.xml data files gives the exact same result as the full syntax. That's what we're fixing here: the full syntax makes use of eval="..." in the xml, while we just keep the "ref()" in the formula with the shortcut syntax.


> > We don't want to support explicit ref() in formulas (a user modifying a formula will anyway not know the xmlid to call).
> 
> Would they really be more cognisant of the raw database id to call?

v9-15 were working this way; nobody ever complained. Moreover, actual users of the db will not have access to xmlids anyway, so, yes, they will be more cognizant of the database ids, as they are in the URLs. I'm not saying this is user-friendly, but it's a really advanced thing to do.

The fact we allow the ref() in the formula was kept from a first  dev version of the v16 reporting, but fp didn't like it, so we went back to doing "eval" in the XML files. Removing it (in the enterprise branch) is done for consistency: ids everywhere is clearer than having sometimes ids and sometimes explicit "ref(...)" for the user.

**Review by smetl — APPROVED** [review]

**Comment by oco-odoo** [comment]
@odoo/rd-security Could any of you unblock this ? We'd like it merged today; we need it before OXP. Thanks in advance :)

**Comment by xmo-odoo** [comment]
> Well, that's why you don't see why enterprise isn't enough. Maybe ask this first ?

... I'm asking both at the same time? I can't know what I don't know, or the unknown dependencies between the things I don't know, so I can only point out what confuses me or what I don't understand?

> Anyway, I can explain, but don't complain if the commit message considers you know the basic models used in account_reports...

:shrug: 

> Shortcuts are just a shorter way to create the same expressions. So, it would make sense that loading them from the *.xml data files gives the exact same result as the full syntax. That's what we're fixing here: the full syntax makes use of eval="..." in the xml, while we just keep the "ref()" in the formula with the shortcut syntax.

The problem with that assertion is that one is applied during module loading to static data, while the other is runtime execution on arbitrary data. You do understand how the trust context and security perimeter are completely different yes?

Especially when using one of the biggest footguns possible as a shortcut in the implementation of a shortcut? Because if all you need to do is resolve `ref` expressions you don't need `safe_eval`, you can -- and should -- do it with `re.sub`, something along the lines of:
```python
re.sub(
    r'''\bref\((?P<quot>['"])(?P<xid>.+?)(?P=quot)\)''',
    lambda m: str(self.env.ref(m['xid'])),
    formula
)
````

**Comment by smetl** [comment]
@robodoo r+

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**
